### DISCUSSION TITLE
Entities vs Packages

### DISCUSSION THREAD

**Discussion by ricochet** [bytecodealliance/SIG-Registries#21] [Polls]
While working on adding a [glossary](https://github.com/bytecodealliance/SIG-Registries/pull/6), I noticed that we're using two terms interchangeably, "entitities" and "packages". 

For examples of how these two terms are used, see the "entity-types" field in the [Publish API](https://github.com/bytecodealliance/SIG-Registries/blob/main/proposals/Publish-API.md#create-unpublished-release) and in text descriptions for the [discovery API proposal](https://github.com/bytecodealliance/SIG-Registries/pull/16/files). 

Earlier docs refer to these as packages:

- [Registry Requirements](https://docs.google.com/document/d/1jv4Vh9o4LNT_XV9sklY1N840vElSI_dilnenzuwkraM/edit#heading=h.s6m06pefqgfb) defines package types.
- [Registry Semantics sketch](https://docs.google.com/document/d/1FxSuSYL0LkGb2jueUAcUu-h4sKEsdiWgMd4axsR95F8/edit#heading=h.vmgdddmg91yf)
- [Log-based Package Registry Sketch](https://docs.google.com/document/d/1X8vBHNkB48XbnRsnX2_2L45a8Hfkk-FEalCFmi3jvCE/edit)

The purpose of this discussion is to agree and commit to one term. I don't have a strong preference either way, but I do see a reasonable argument for one over the other.

While we seem to have adopted a very generic term, entities, to make it clear that many different things of various types are managed by the registry and have been sticking to this term closely in the API proposals added to this repo, colloquially we consistently refer to these as packages when speaking about them in meetings. I've seen many different APIs grow over the years and it's the colloquial term that is always the most pervasive and will inevitably be fed back into the project or in projects that build on a given API.

I don't see issues with changing "entities" and "entity types" to "packages" and "package types" in our current proposals, but if there is one, please call this out.

**Comment by kulakowski-wasm** [comment]
My braindump (and tl;dr I like *package* a lot more):

The choice of *entity* seems to capture a desire to acknowledge that different kinds of things are published to and fetched from registries. Is this actually the case? Was it something else? I think this term came just as I was coming into this group.

*package*, to me, captures a lot of the essence of what objects a registry operates on. There's metadata and variously structured contents (more or less) hermetically sealed as a well-identifiable unit. I feel that there will be operations in common for different types of packages and that they share some semantic content (a profile has a publisher just as much as a component, say).

The latter of these pulls a lot more heavily with me.


I do have a set of questions maybe for @lukewagner since I think the `Registry Semantics sketch` doc is his.

That doc describes a registry as a collection of namespaces, imports, and packages, but doesn't really give any meaning to the namespace and import parts.

What do namespaces and imports do? Just effect eg resolution?

Are they distinct from packages in the sense that they really are totally different kinds of objects, operated on by different agents, having very different metadata requirements from packages/entities?

**Comment by lukewagner** [comment]
I'm happy to leave namespaces/imports out of the current glossary, since they were just in the sketch doc to sketch out some ideas, not permanently establish the terminology and only add them once we have a more concrete HTTP API proposal.

**Reply by kulakowski-wasm** [reply]
Sure! To be clear, maybe a better version of the question was "are namespaces and imports (or the imagined future design pressures they are standing for ;p) the historical reason for saying `entity` instead of `package`", and I take it the answer to that is "no".

**Reply by lukewagner** [reply]
Oh, hah, gotcha; agreed.
